src/Nix/Value.hs

```diff
@@ -622,40 +640,43 @@ data ValueType
 
 -- | Determine type of a value
 valueType :: NValueF a m r -> ValueType
-valueType = \case
-  NVConstantF a -> case a of
-    NURI   _ -> TString NoContext
-    NInt   _ -> TInt
-    NFloat _ -> TFloat
-    NBool  _ -> TBool
-    NNull    -> TNull
-  NVStrF ns  ->
-    TString $
-      bool
-        NoContext
-        HasContext
-        $ stringHasContext ns
-  NVListF{}    -> TList
-  NVSetF{}     -> TSet
-  NVClosureF{} -> TClosure
-  NVPathF{}    -> TPath
-  NVBuiltinF{} -> TBuiltin
+valueType =
+  \case
+    NVConstantF a ->
+      case a of
+        NURI   _ -> TString NoContext
+        NInt   _ -> TInt
+        NFloat _ -> TFloat
+        NBool  _ -> TBool
+        NNull    -> TNull
+    NVStrF ns  ->
+      TString $
+        bool
+          NoContext
+          HasContext
+          (stringHasContext ns)
+    NVListF{}    -> TList
+    NVSetF{}     -> TSet
+    NVClosureF{} -> TClosure
+    NVPathF{}    -> TPath
+    NVBuiltinF{} -> TBuiltin
 
 
 -- | Describe type value
 describeValue :: ValueType -> String
-describeValue = \case
-  TInt               -> "an integer"
-  TFloat             -> "a float"
-  TBool              -> "a boolean"
-  TNull              -> "a null"
-  TString NoContext  -> "a string"
-  TString HasContext -> "a string with context"
-  TList              -> "a list"
-  TSet               -> "an attr set"
-  TClosure           -> "a function"
-  TPath              -> "a path"
-  TBuiltin           -> "a builtin function"
+describeValue =
+  \case
```

Comment:
I've never seen this style. It would seem to be extra indent for no purpose?

Reply:
The lispy-type of code is even more powerful to refactor because of the Structured Haskell Mode.

Structured programming is Lisp-born and is powerful in it, because, as known, of ap parens reasons.

But in my case, the Structured Haskell Mode does not work in my Emacs (VI bindings have a side-car binding system, SHM uses native binding system).

Reply:
I will have to finish reading this later. I do have a soft spot for "Structured Haskell Mode". I would not recommend replacing pattern matching with `maybe` and `bool` though in large cases, however. Even Lisp came towards pattern matching with https://docs.racket-lang.org/guide/syntax-case.html and https://www.gnu.org/software/guile/manual/html_node/Syntax-Case.html. https://www.pyret.org/ led by one of the https://htdp.org/ authors also makes pattern matching part of the language.

Bottom line is yes there are things to learn from lisp and structured editing is very good, but I would not make avoiding pattern matching one of those lessons.

Reply:
It all started with someone who cares suggesting to try to use `maybe` and stick to it to open a forgotten realm.

Reply:
Sorry, maybe it was a bit direct rambling.

> I've never seen this style. It would seem to be extra indent for no purpose?

They serve the purpose.
All the heads of the functions are at the beginning of the lines. And as such inclusion of 1 function into another - shown by the indentation.

Code reads diagnonally from down right to the up left, the reader at once sees all the functions and what they carrie, eye movement ends at the function definition name.

That is the purpose of the indentation.

I also started to like indenting the `do` sometimes, to diagonally outline the `do` blocks inside the function body by indentation.

`\case` is a lambda function, with `\case` being the head of the function, and as such - gets indented.
So diagonal indentations have a purpose.

Haskell being the abstract terse language can bear the diagonal alignment of code. If the Haskell function does not fit on the screen it needs to be refactored, if that is not just a large pattern match to constructors.

So those `\case`, when they are more then a couple of lines - are just a nod as a placeholder waiting to be properly named local function, waiting to be refactored.
It is a style that enables refactoring in the code.

Reply:
Good example. how in the head to parse this:
```haskell
mapAttrs_ f xs = fromValue @(AttrSet (NValue t f m)) xs >>= \aset -> do
  let pairs = M.toList aset
  values <- for pairs $ \(key, value) ->
    defer @(NValue t f m)
      $   withFrame Debug (ErrorCall "While applying f in mapAttrs:\n")
      $   callFunc ?? value
      =<< callFunc f (nvStr (makeNixStringWithoutContext key))
  toValue . M.fromList . zip (fmap fst pairs) $ values
```
???
Is this better?

How
```haskell
defer @(NValue t f m)
      $   withFrame Debug (ErrorCall "While applying f in mapAttrs:\n")
      $   callFunc ?? value
      =<< callFunc f (nvStr (makeNixStringWithoutContext key))
```
works.
How the `$` `??` `=<<` map where?

To understand it - the person would start sectioning the code with parens (aka splitting into lambdas) - breaking it down into Lisp essentially.

Reply:
```haskell
mapAttrs_ f xs =
  do
    aset <- fromValue @(AttrSet (NValue t f m)) xs
    let
      pairs = M.toList aset

    values <-
      (traverse
        (\ (key, value) ->
          (defer @(NValue t f m)
            (withFrame Debug (ErrorCall "While applying f in mapAttrs:\n")
            ((=<<)
              (\ x -> (callFunc ?? value) x)
              (callFunc f (nvStr (makeNixStringWithoutContext key)))))))
        pairs
      )
    (toValue . M.fromList . zip (fmap fst pairs)) values
```

It is:
```haskell
mapAttrs_ f xs =
  do
    attrset <- fromValue @(AttrSet (NValue t f m)) xs
    let
      keyVals = M.toList attrset

    newVals <-
      traverse
        (\ (key, val) ->
          defer @(NValue t f m)
            (withFrame Debug (ErrorCall "While applying f in mapAttrs:\n")
            (do
              runFunForKey <- callFunc f $ nvStr $ makeNixStringWithoutContext key
              callFunc runFunForKey val
              )))
        keyVals
    (toValue . M.fromList . zip (fst <$> keyVals)) newVals
```

Of course, it also can be:
```haskell
mapAttrs_ f xs = do
    nixAttrset <- fromValue @(AttrSet (NValue t f m)) xs
    let
      keyVals = M.toList nixAttrset
      applyFunToKeyVal (key, val) = do
        runFunForKey <- callFunc f $ nvStr $ makeNixStringWithoutContext key
        callFunc runFunForKey val
    newVals <- traverse (defer @(NValue t f m) .
      withFrame Debug (ErrorCall "While applying f in mapAttrs:\n") . applyFunToKeyVal
      ) keyVals
    toValue $ M.fromList $ zip (fst <$> keyVals) newVals
```

This a resulting code with the empty lines removed.

But next tend to read better:
```haskell
mapAttrs_ f xs =
  do
    nixAttrset <- fromValue @(AttrSet (NValue t f m)) xs
    let
      keyVals = M.toList nixAttrset
      keys = fst <$> keyVals

      applyFunToKeyVal (key, val) =
        do
          runFunForKey <- callFunc f $ nvStr $ makeNixStringWithoutContext key
          callFunc runFunForKey val

    newVals <-
      traverse
        (defer @(NValue t f m) . withFrame Debug (ErrorCall "While applying f in mapAttrs:\n") . applyFunToKeyVal)
        keyVals

    toValue $ M.fromList $ zip keys newVals
```

There are scientific papers that determined that the most important alignment property in readability is the amount of space between lines.
People understand the text better when there is more space between lines, it is psychologically easier to read the text.

The best open sum-up material that I found on it so far is: https://ajet.org.au/index.php/AJET/article/view/2286/1117. It is an old studied scientific topic, so there are no open new papers. In there is a number of references to the interconnection between readability, typography, and linespace. In short, good linespace helps readability a lot.

Reply:
> I will have to finish reading this later. I do have a soft spot for "Structured Haskell Mode". I would not recommend replacing pattern matching with maybe and bool though in large cases, however. Even Lisp came towards pattern matching...

Of course. `maybe` and such can not replace complex pattern matches, `bool` `maybe` `either` binary, while pattern matches are multicast.
Also Haskell pattern matches go together with guards, ViewPatterns and such others.

I so far not once replaced the complex pattern match in the code, you can check 8). There is only 1 very minor `null` dup sin I just contributed, but the dup there replicates the upstream Nix bug behavior.

Reply:
+===

Well, I would retreat again.

I looked at the line marked in the review message, and at the message. And knowing the style I wrote in, started to discuss it. But not looked what the code is beneath the line marked. Alignment of those case statements, yes, alignment here was a bit vacuous.
Just during the work process aligned these also, if that is possible to allow by the align rules tend to favor giving more align-right to the biggish functions and the biggish pattern matches,  to indicate to the eyes that they are not at the beginning of the line and that they are in fact a block of code inside something that is up the screen.

But that is a minor thing for me, maybe would try to obstrain more from that, also can revert these.